### PR TITLE
[pkg/ottl] Change context inferrer to use functions and enums as hints

### PR DESCRIPTION
**Pull request by edmocosta**
#### Description

This PR is part of https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/29017, and a spin-off from https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/36820. It changes the existing context inferrer logic to also take into consideration the functions and enums used on the statements. (https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/36820#issuecomment-2542028806)

New logic:
- Find all `path`, function names(`editor`, `converter`), and `enumSymbol` on the statements
- Pick the highest priority context (same existing logic) based on the `path.Context` values
- If the chosen context does not support all used functions and enums, it goes through it's lower contexts (wide scope contexts that does support the chosen context as a path context) testing them and choosing the first one that supports them.
- If no context that can handle the functions and enums is found, the inference fail and an empty value is returned. 

The parser collection was adapted to support the new context inferrer configuration requirements. 

**Other important changes:**

Currently, it's possible to have paths to contexts root objects, for example: `set(attributes["body"], resource)`. Given  `resource` has no dot separators on the path, the grammar extracts it into the `path.Fields` slice, letting the `path.Context` value empty. Why? This grammar behaviour is still necessary to keep backward compatibility with paths without context, otherwise it would start requiring contexts for all paths independently of the parser configuration.

Previous PRs didn't take this edge case into consideration, and a few places needed to be changed to address it:
 - Context inferrer (`getContextCandidate`)
 - Parser `prependContextToStatementPaths` function.
 - Reusable OTTL contexts (`contexts/internal`) (not part of this PR, it will be fixed by https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/36820)

When/If we reach the point to deprecate paths _without_ context, all those conditions can be removed, and the grammar changed to require and extract the `path` context properly.
 

#### Link to tracking issue
https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/29017

#### Testing
Unit tests

#### Documentation
No changes

### CODE REVIEW

**Comment by TylerHelmuth** [comment]
> Currently, it's possible to have paths to contexts root objects, for example: set(attributes["body"], resource). Given resource has no dot separators on the path, the grammar extracts it into the path.Fields slice, letting the path.Context value empty. Why? This grammar behaviour is still necessary to keep backward compatibility with paths without context, otherwise it would start requiring contexts for all paths independently of the parser configuration.

ya all the context implementations allowed these paths, but the grammar didn't actually allow you to use them in any statements I don't think. I don't think we need to support a path referencing the entire object higher in the otlp payload. I am ok with removing this capability from our context implementations and making statements like `set(attributes["body"], resource)` fail.

File: pkg/ottl/functions.go [hunk 1]
```diff
@@ -103,13 +103,30 @@ func (p *Parser[K]) parsePathContext(path *path) (string, []field, error) {
...
 	}
 
 	return "", path.Fields, nil
 }
 
+// When a path has no dots separators (e.g.: resource), the grammar extracts it into the
+// path.Fields slice, letting the path.Context empty. This function verifies if a given
+// path is accessing any configured (ottl.WithPathContextNames) object root, returning
+// true and the resolved context name.
+func (p *Parser[K]) isPathToContextRootData(pat *path) (bool, string) {
```
Comment (TylerHelmuth) [inline pkg/ottl/functions.go:120]: If we got rid of support for `set(attributes["body"], resource)` and similar paths separately by removing path definitions like https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/a70fa09f8e31f6847cc24f0e8d16b1e78a69c1b6/pkg/ottl/contexts/internal/resource.go#L21, could we get rid of this logic?

File: pkg/ottl/context_inferrer.go [hunk 2]
```diff
@@ -3,75 +3,254 @@
...
+				inferredContext = candidate
+				inferredContextPriority = candidatePriority
 			}
 		}
+		for function := range statementFunctions {
+			requiredFunctions[function] = struct{}{}
+		}
+		for enum := range statementEnums {
+			requiredEnums[enum] = struct{}{}
+		}
```
Comment (TylerHelmuth) [inline pkg/ottl/context_inferrer.go:106]: Can `s.getParsedStatementHints(parsed)` return `map[string]struct{}{}` and `map[enumSymbol]struct{}{}` so we don't need to do this data struct translation?

File: pkg/ottl/functions.go (hunk 1 above)
Comment (TylerHelmuth) [reply pkg/ottl/functions.go:120]: Created https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/36872

Comment (edmocosta) [reply pkg/ottl/functions.go:120]: That's nice, removing support to it simplifies the code a lot, thanks! I'll remove this logic.

**Review by TylerHelmuth — APPROVED** [review]

**Comment by TylerHelmuth** [comment]
@evan-bradley since everything is internal and can be refactored later if needed im gonna merge now to keep things moving.

**Pull request merged**